### ISSUE TITLE
Override layer parameters on predefined states (Hover, Pressed, etc)

### ISSUE THREAD

**Issue by GuillaumeSalles** [airbnb/Lona#52]
## Summary of Idea

In order to create more interactive components, I propose to let the user override layer parameters on predefined states like `Hover`, `Pressed`, etc...

The material button is a good example of a component that would benefit from such flexibility. https://material.io/guidelines/components/buttons.html

## Initial Design

Add a dropdown at the top (or bottom) of the layer parameters sidebar with all states allowed based on the layer type. 
The dropdown default value will be `Default`. 
If the user select a state other than `Default`, the layer parameters can be overridden for the selected state.
In the canvas, the component is rendered based on the layers selected state.

## Proposed Implementation

The states are limited based on the layer type. (A `Disable` state does not make sense for a `Text` layer.)

To limit the scope of this proposal, only the `Hover` and `Pressed` states could be supported at first.

The transition between two parameter values when a state changes is not covered in this proposal. (No easing function for now, just instant transitions.) 

### Component spec updates

|Attribute|Type|Required|Description|
|---|---|---|---|
|`hover`|`JSON`|No|Override parameters values when the mouse go over the layer|
|`pressed`|`JSON`|No|Override parameters values when the layer is pressed with a finger, a mouse, or another input device|

### Cons

- This solution might be too flexible and could make the code generation too complex.
- It matches the css way of handling visual states, however, I don't know if it can be done in native land. 

An alternative solution could be to add only specific properties like `pressedBackground`, `hoverColor`, but I think the number of properties will explode very fast if we don't want to limit the designer creativity.

**Comment by dabbott** [comment]
Thanks for getting this started @GuillaumeSalles! This is a great writeup. I completely agree that we should figure out interactivity. 

## Thoughts on this approach

I like this approach because:
- It's very easy and clear to imagine how I would use it. Conceptually it's similar to how I think about states when I design. There's a pressed state, a hover state, etc, and for each state I can specify a new set of parameters.

Simplicity is a _huge_ advantage that might outweigh any cons.

(In the following feedback, I'm making the assumption that custom components can have pressed and hover states too)

I think you've also identified the main cons here - possible code generation complexity, especially on native. To add a little more detail:
- On both web and native, we would also have to think about _which_ element will actually be handling the hover and press events. E.g. if you have nested custom components, we'll have to find e.g. the first `div` child and set event handlers on that, since custom React components don't have DOM event handlers. On native, some components e.g. Text might need to be wrapped in something that can handle events.
- If the set of states we support expands, then every component would potentially need to show a pretty large list of possible states, most of which wouldn't be used.
- We'll likely have components that aren't easily modeled as finite state machines, especially as we add more kinds of states. E.g. we might need to express states like `pressed && hover`, `pressed && !hover`, `!hover && pressed`. It think it would get pretty cumbersome to express this via a state picker UI, and instead we should leverage the Logic for this sort of thing.

I have a few ideas. I don't think these are necessarily mutually exclusive with the approach you've outlined -- potentially we can combine the best of a few different ideas.

My goal is to come up with a solid foundation for handling interactions and states in a flexible and scalable way that also outputs reasonable code.

## Idea 1: `Button` Component

We have a component that very explicitly generates divs/Views that can handle press and hover events. This limits the flexibility. We don't have to consider generating event handling code on View/Text/Image/Custom, just Button.

The Button component would then expose `pressed` and `hover` _somehow_ (not quite sure what's best here) for use in Logic. It could be something like:

```
If layers->MyButton->state->pressed
  Assign custom Color `Red 500` to layers->MyView->backgroundColor 
```

Or possibly

```
When layers->MyButton->state->pressed changes
  If newValue
    Assign custom Color `Red 500` to layers->MyView->backgroundColor 
```

### Pros

- Simplifies code generation by limiting event handling to a specific component
- We can have different components for different kinds of events (e.g. Draggable) so we don't have to have a growing list of events that can be applied to every component. Custom components won't need to support e.g. pressed/hover by default
- `pressed` and `hover` can be used in arbitrary ways in Logic. E.g. `pressed && !hover`. I can also have a single component respond to the states of multiple children: `Button1.state.pressed && !Button2.state.hover` (pressed and hover aren't great examples of this, since that doesn't make much sense... but when we start thinking about states more broadly, I think it'll be relevant)

### Cons

- The mental model for defining states in logic isn't as easy/clear as with a state picker dropdown and setting layer parameters
- How would we let you easily design these? It's easy to imagine how clicking a Button would set the pressed state, but you can't both click the button and tweak it's parameters. You would have to go back and forth, which would not be ideal. Maybe we can solve this with...

## Idea 2: State

This is similar to React state. We would have another tab at the bottom for `state` (essentially the same as the `parameters` tab). We would have a way to `setState` using Logic. We would have a UI for overriding the state manually, in order to more easily design for a particular state without having to e.g. click the button to enter the pressed state.

The main difference from React state is that I think we'd want to expose states (_readonly_) to parent components. In terms of implementation, we would wire up the function props automatically.

```
Assign custom Boolean false to state->active
If layers->Button1->state->hovered
  If layers->Button2->state->hovered
    Assign custom Boolean true to state->active
```

If the Logic were like this, we'd be mixing internal state and propagating changes to the parent component. When we generate code for React, we would create implicit change handler props for each value in state. E.g. if we had an `active` state value, we would generate `onChangeActive` to notify parent components. We would then write `Assign custom Boolean true to state->active` as `setState({ active: true }); onChangeActive(true);`. The parent would have to duplicate some of the state of its children.

If we wanted to be more explicit about duplicating state into the parent, the Logic might look like:

```
When layers->Button1->state->hovered changes
  Assign newValue to state->button1Active
When layers->Button2->state->hovered changes
  Assign newValue to state->button2Active
Assign custom Boolean false to state->active
If state->button1Active
  If layers->button2Active
    Assign custom Boolean true to state->active
```

On native (Swift at least), code generation is pretty straightforward I think. We create public getters/setters for each value in state.

### Pros

- This gives us a solid foundation for arbitrary state handling. States aren't implicit to specific interactive components, but rather a new concept that we can use in custom components too.
- States can be overridden using the UI to allow easily designing for specific states

### Cons

- Complexity. We're adding yet another concept to an already complicated tool

Alternately, we could just use parameters for this, but then we're really overloading what parameters do. Even in the examples I just outlined, state is already somewhat overloaded (it both stores internal state and is used by parent components).

## In summary

I think if we combined `Button` components and `state`, we'd have a robust system for handling interactive states via logic. However, the complexity is pretty high, and writing logic is generally slower and less visual than making the equivalent change using the GUI.

The initial approach you outlined is easy to use, since it maps well the mental model of designing for different states. However, it doesn't scale as well to new or more complex use cases and it can lead to some tricky code generation.

I think we should try to find the best balance of simplicity **and** a solid foundation for more kinds of use cases.

That's a lot of stuff but... thoughts?
